source/structuring.rst

```diff
@@ -14,402 +14,545 @@ functions is better than copying-and-pasting duplicate bits of code
+   single: $ref
+   single: structuring; recursion; $ref
 
-      "definitions": {
-        "address": {
-          *"$id": "#address",
-          "type": "object",
-          "properties": {
-            "street_address": { "type": "string" },
-            "city":           { "type": "string" },
-            "state":          { "type": "string" }
-          },
-          "required": ["street_address", "city", "state"]
-        }
-      },
+.. _recursion:
 
-      "type": "object",
+Recursion
+---------
 
+The ``$ref`` keyword may be used to create recursive schemas that refer
+to themselves. For example, you might have a ``person`` schema that
+has an array of ``children``, each of which are also ``person``
+instances.
+
+.. schema_example::
+
+    {
+      "type": "object",
       "properties": {
-        *"billing_address": { "$ref": "#address" },
-        *"shipping_address": { "$ref": "#address" }
+        "name": { "type": "string" },
+        "children": {
+          "type": "array",
+    *      "items": { "$ref": "#" }
+        }
       }
     }
+    --
+    // A snippet of the British royal family tree
+    {
+      "name": "Elizabeth",
+      "children": [
+        {
+          "name": "Charles",
+          "children": [
+            {
+              "name": "William",
+              "children": [
+                { "name": "George" },
+                { "name": "Charlotte" }
+              ]
+            },
+            {
+              "name": "Harry"
+            }
+          ]
+        }
+      ]
+    }
 
-.. note::
+Above, we created a schema that refers to itself, effectively creating
+a "loop" in the validator, which is both allowed and useful. Note,
+however, that a ``$ref`` referring to another ``$ref`` could cause
+an infinite loop in the resolver, and is explicitly disallowed.
 
-    This functionality isn't currently supported by the Python ``jsonschema``
-    library.
+.. schema_example::
 
-Extending
----------
+    {
+      "definitions": {
+        "alice": { "$ref": "#/definitions/bob" },
+        "bob": { "$ref": "#/definitions/alice" }
+      }
+    }
 
-The power of ``$ref`` really shines when it is used with the
-combining keywords ``allOf``, ``anyOf`` and ``oneOf`` (see
-:ref:`combining`).
+.. index::
+   single: bundling
+   single: $id
+   single: structuring; bundling; $id
 
-Let's say that for a shipping address, we want to know whether the
-address is a residential or business address, because the shipping
-method used may depend on that.  For a billing address, we don't
-want to store that information, because it's not applicable.
+.. _bundling:
 
-To handle this, we'll update our definition of shipping address::
+Bundling
+--------
 
-    "shipping_address": { "$ref": "#/definitions/address" }
+Working with multiple schema documents is convenient for development,
+but it is often more convenient for distribution to bundle all of your
+schemas into a single schema document. This can be done using the
```

Comment:
```suggestion
schemas into a single schema document (Compound Schema Document). This can be done using the
```

Reply:
Remember this is for draft-07. The concept of a "Compound Schema" didn't exist yet. In draft 2019-09, `$id` indicates an embedded schema, but before that, `$id` only changes the base URI. It's still considered a single schema document.